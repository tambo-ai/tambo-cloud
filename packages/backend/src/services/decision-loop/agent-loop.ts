import { Message } from "@ag-ui/core";
import {
  LegacyComponentDecision,
  MessageRole,
  ThreadMessage,
  ToolCallRequest,
} from "@tambo-ai-cloud/core";
import OpenAI from "openai";
import { threadMessagesToChatCompletionMessageParam } from "../../util/thread-message-conversion";
import { AgentClient } from "../llm/agent-client";

export async function* runAgentLoop(
  agentClient: AgentClient,
  messages: ThreadMessage[],
  strictTools: OpenAI.Chat.Completions.ChatCompletionTool[],
  //   customInstructions: string | undefined,
): AsyncIterableIterator<LegacyComponentDecision> {
  const chatCompletionMessages =
    threadMessagesToChatCompletionMessageParam(messages);
  // const systemPromptArgs = customInstructions
  //     ? { custom_instructions: customInstructions }
  //     : {};

  const stream = agentClient.streamRunAgent({
    messages: chatCompletionMessages,
    tools: strictTools,
    // promptTemplateName: "decision-loop",
    // promptTemplateParams: {
    //   chat_history: chatCompletionMessages,
    //   ...systemPromptArgs,
    // },
  });
  for await (const event of stream) {
    const { message } = event;
    const toolCallId = getToolCallId(message);
    const toolCallRequest = getToolCallRequest(message);
    yield {
      id: message.id,
      role: message.role as MessageRole,
      message: message.content || "",
      componentName: null,
      props: null,
      componentState: null,
      reasoning: "",
      statusMessage: "",
      completionStatusMessage: "",
      toolCallRequest: toolCallRequest,
      toolCallId: toolCallId,
    };
  }
}
function getToolCallId(message: Message) {
  if (message.role === "assistant") {
    return message.toolCalls?.[0]?.id;
  }
  if (message.role === "tool") {
    return message.toolCallId;
  }
  return undefined;
}

function getToolCallRequest(message: Message): ToolCallRequest | undefined {
  if (message.role !== "assistant" || !message.toolCalls?.length) {
    return;
  }
  const toolCall = message.toolCalls[0];
  try {
    const parameters: Record<string, unknown> = JSON.parse(
      toolCall.function.arguments ? toolCall.function.arguments : "{}",
    );
    return {
      toolName: toolCall.function.name,
      parameters: Object.entries(parameters).map(([key, value]) => ({
        parameterName: key,
        parameterValue: value,
      })),
    };
  } catch (e) {
    console.warn(
      `Error parsing tool call arguments for tool '${toolCall.function.name}'`,
      e,
      toolCall.function.arguments,
    );
    return;
  }
}
