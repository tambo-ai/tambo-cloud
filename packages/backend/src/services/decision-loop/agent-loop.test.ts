import {
  LegacyComponentDecision,
  MessageRole,
  ThreadMessage,
} from "@tambo-ai-cloud/core";
import OpenAI from "openai";
import {
  AgentClient,
  AgentMessage,
  AgentResponse,
  AgentResponseType,
} from "../llm/agent-client";
import { runAgentLoop } from "./agent-loop";

// Mock the agent client
const createMockAgentClient = (): AgentClient => {
  return {
    streamRunAgent: jest.fn(),
  } as unknown as AgentClient;
};

// Helper to create mock events
const createMockEvent = (message: AgentMessage): AgentResponse => ({
  type: AgentResponseType.MESSAGE,
  message,
});

// Helper to create base thread message
const createBaseThreadMessage = (): Omit<
  ThreadMessage,
  "id" | "role" | "content"
> => ({
  threadId: "test-thread",
  componentState: {},
  createdAt: new Date("2024-01-01T00:00:00Z"),
});

// Helper to create mock tools
const createMockTools = (): OpenAI.Chat.Completions.ChatCompletionTool[] => [
  {
    type: "function",
    function: {
      name: "test_tool",
      description: "A test tool",
      parameters: {
        type: "object",
        properties: {
          param1: { type: "string" },
          param2: { type: "number" },
        },
      },
    },
  },
];

describe("runAgentLoop", () => {
  let mockAgentClient: AgentClient;
  let mockStream: AsyncIterableIterator<AgentResponse>;

  beforeEach(() => {
    mockAgentClient = createMockAgentClient();
    mockStream = {
      [Symbol.asyncIterator]: jest.fn(),
    } as unknown as AsyncIterableIterator<AgentResponse>;
    jest.mocked(mockAgentClient.streamRunAgent).mockReturnValue(mockStream);
  });

  describe("message event sequences", () => {
    it("should handle complete assistant message sequence", async () => {
      const messages: ThreadMessage[] = [
        {
          ...createBaseThreadMessage(),
          id: "user-1",
          role: MessageRole.User,
          content: [{ type: "text", text: "Hello, how are you?" }],
        },
      ];

      const events = [
        createMockEvent({
          id: "assistant-1",
          role: "assistant",
          content: undefined,
          reasoning: ["I need to respond to the user's greeting"],
        }),
        createMockEvent({
          id: "assistant-1",
          role: "assistant",
          content: "Hello! I'm doing well, thank you for asking. ",
          reasoning: ["I should be friendly and respond to their greeting"],
        }),
        createMockEvent({
          id: "assistant-1",
          role: "assistant",
          content:
            "Hello! I'm doing well, thank you for asking. How can I help you today?",
          reasoning: ["I should also offer to help them"],
        }),
      ];

      // Mock the async iterator
      let eventIndex = 0;
      jest.mocked(mockStream[Symbol.asyncIterator]).mockReturnValue({
        next: async () => {
          if (eventIndex < events.length) {
            return { value: events[eventIndex++], done: false };
          }
          return { done: true, value: undefined };
        },
      } as AsyncIterableIterator<AgentResponse>);

      const results: LegacyComponentDecision[] = [];
      for await (const result of runAgentLoop(
        mockAgentClient,
        messages,
        createMockTools(),
      )) {
        results.push(result);
      }

      expect(results).toMatchSnapshot();
    });

    it("should handle user message sequence", async () => {
      const messages: ThreadMessage[] = [];

      const events = [
        createMockEvent({
          id: "user-1",
          role: "user",
          content: "",
          reasoning: [],
        }),
        createMockEvent({
          id: "user-1",
          role: "user",
          content: "What's the weather like?",
          reasoning: [],
        }),
      ];

      let eventIndex = 0;
      jest.mocked(mockStream[Symbol.asyncIterator]).mockReturnValue({
        next: async () => {
          if (eventIndex < events.length) {
            return { value: events[eventIndex++], done: false };
          }
          return { done: true, value: undefined };
        },
      } as AsyncIterableIterator<AgentResponse>);

      const results: LegacyComponentDecision[] = [];
      for await (const result of runAgentLoop(
        mockAgentClient,
        messages,
        createMockTools(),
      )) {
        results.push(result);
      }

      expect(results).toMatchSnapshot();
    });
  });

  describe("tool call event sequences", () => {
    it("should handle complete tool call sequence", async () => {
      const messages: ThreadMessage[] = [];

      const events = [
        createMockEvent({
          id: "assistant-1",
          role: "assistant",
          content: "I'll help you with that.",
          reasoning: ["I should use a tool to get the information"],
        }),
        createMockEvent({
          id: "assistant-1",
          role: "assistant",
          content: "I'll help you with that.",
          toolCalls: [
            {
              id: "call_123",
              type: "function" as const,
              function: {
                name: "test_tool",
                arguments: "",
              },
            },
          ],
          reasoning: ["I need to call the test_tool"],
        }),
        createMockEvent({
          id: "assistant-1",
          role: "assistant",
          content: "I'll help you with that.",
          toolCalls: [
            {
              id: "call_123",
              type: "function" as const,
              function: {
                name: "test_tool",
                arguments: '{"param1": "value1", "param2": 42}',
              },
            },
          ],
          reasoning: ["I've prepared the tool call with parameters"],
        }),
      ];

      let eventIndex = 0;
      jest.mocked(mockStream[Symbol.asyncIterator]).mockReturnValue({
        next: async () => {
          if (eventIndex < events.length) {
            return { value: events[eventIndex++], done: false };
          }
          return { done: true, value: undefined };
        },
      } as AsyncIterableIterator<AgentResponse>);

      const results: LegacyComponentDecision[] = [];
      for await (const result of runAgentLoop(
        mockAgentClient,
        messages,
        createMockTools(),
      )) {
        results.push(result);
      }

      expect(results).toMatchSnapshot();
    });

    it("should handle tool call with invalid JSON arguments", async () => {
      const messages: ThreadMessage[] = [];

      const events = [
        createMockEvent({
          id: "assistant-1",
          role: "assistant",
          content: "I'll call a tool.",
          toolCalls: [
            {
              id: "call_123",
              type: "function" as const,
              function: {
                name: "test_tool",
                arguments: '{"invalid": json}', // Invalid JSON
              },
            },
          ],
          reasoning: ["I need to call the test_tool"],
        }),
      ];

      let eventIndex = 0;
      jest.mocked(mockStream[Symbol.asyncIterator]).mockReturnValue({
        next: async () => {
          if (eventIndex < events.length) {
            return { value: events[eventIndex++], done: false };
          }
          return { done: true, value: undefined };
        },
      } as AsyncIterableIterator<AgentResponse>);

      // Mock console.warn to avoid test output
      const consoleSpy = jest.spyOn(console, "warn").mockImplementation();

      const results: LegacyComponentDecision[] = [];
      for await (const result of runAgentLoop(
        mockAgentClient,
        messages,
        createMockTools(),
      )) {
        results.push(result);
      }

      expect(consoleSpy).toHaveBeenCalledWith(
        expect.stringContaining("Error parsing tool call arguments"),
        expect.any(Error),
        '{"invalid": json}',
      );
      expect(results).toMatchSnapshot();

      consoleSpy.mockRestore();
    });

    it("should handle tool result message", async () => {
      const messages: ThreadMessage[] = [];

      const events = [
        createMockEvent({
          id: "tool-1",
          role: "tool",
          content: "Tool execution result",
          toolCallId: "call_123",
          reasoning: [],
        }),
      ];

      let eventIndex = 0;
      jest.mocked(mockStream[Symbol.asyncIterator]).mockReturnValue({
        next: async () => {
          if (eventIndex < events.length) {
            return { value: events[eventIndex++], done: false };
          }
          return { done: true, value: undefined };
        },
      } as AsyncIterableIterator<AgentResponse>);

      const results: LegacyComponentDecision[] = [];
      for await (const result of runAgentLoop(
        mockAgentClient,
        messages,
        createMockTools(),
      )) {
        results.push(result);
      }

      expect(results).toMatchSnapshot();
    });
  });

  describe("reasoning sequences", () => {
    it("should handle assistant message with reasoning", async () => {
      const messages: ThreadMessage[] = [];

      const events = [
        createMockEvent({
          id: "assistant-1",
          role: "assistant",
          content: undefined,
          reasoning: [
            "The user is asking about the weather",
            "I should check the current weather conditions",
            "I'll need to use a weather tool",
          ],
        }),
        createMockEvent({
          id: "assistant-1",
          role: "assistant",
          content: "Let me check the weather for you.",
          reasoning: [
            "The user is asking about the weather",
            "I should check the current weather conditions",
            "I'll need to use a weather tool",
            "I should let them know I'm looking it up",
          ],
        }),
      ];

      let eventIndex = 0;
      jest.mocked(mockStream[Symbol.asyncIterator]).mockReturnValue({
        next: async () => {
          if (eventIndex < events.length) {
            return { value: events[eventIndex++], done: false };
          }
          return { done: true, value: undefined };
        },
      } as AsyncIterableIterator<AgentResponse>);

      const results: LegacyComponentDecision[] = [];
      for await (const result of runAgentLoop(
        mockAgentClient,
        messages,
        createMockTools(),
      )) {
        results.push(result);
      }

      expect(results).toMatchSnapshot();
    });

    it("should handle reasoning without content", async () => {
      const messages: ThreadMessage[] = [];

      const events = [
        createMockEvent({
          id: "assistant-1",
          role: "assistant",
          content: undefined,
          reasoning: [
            "I need to think about this problem",
            "This is a complex question that requires analysis",
            "Let me break it down step by step",
          ],
        }),
      ];

      let eventIndex = 0;
      jest.mocked(mockStream[Symbol.asyncIterator]).mockReturnValue({
        next: async () => {
          if (eventIndex < events.length) {
            return { value: events[eventIndex++], done: false };
          }
          return { done: true, value: undefined };
        },
      } as AsyncIterableIterator<AgentResponse>);

      const results: LegacyComponentDecision[] = [];
      for await (const result of runAgentLoop(
        mockAgentClient,
        messages,
        createMockTools(),
      )) {
        results.push(result);
      }

      expect(results).toMatchSnapshot();
    });
  });

  describe("complete conversation flow", () => {
    it("should handle full conversation with assistant, tool call, tool result, and follow-up", async () => {
      const messages: ThreadMessage[] = [
        {
          ...createBaseThreadMessage(),
          id: "user-1",
          role: MessageRole.User,
          content: [{ type: "text", text: "What's 2 + 2?" }],
        },
      ];

      const events = [
        // First assistant message
        createMockEvent({
          id: "assistant-1",
          role: "assistant",
          content: undefined,
          reasoning: [
            "The user is asking a math question",
            "I should use a calculator tool",
          ],
        }),
        createMockEvent({
          id: "assistant-1",
          role: "assistant",
          content: "I'll calculate that for you.",
          reasoning: [
            "The user is asking a math question",
            "I should use a calculator tool",
          ],
        }),

        // Tool call
        createMockEvent({
          id: "assistant-1",
          role: "assistant",
          content: "I'll calculate that for you.",
          toolCalls: [
            {
              id: "call_calc_1",
              type: "function" as const,
              function: {
                name: "calculator",
                arguments: '{"expression": "2 + 2"}',
              },
            },
          ],
          reasoning: ["I need to call the calculator tool with the expression"],
        }),

        // Tool result
        createMockEvent({
          id: "tool-1",
          role: "tool",
          content: "4",
          toolCallId: "call_calc_1",
          reasoning: [],
        }),

        // User follow-up
        createMockEvent({
          id: "user-2",
          role: "user",
          content: "What about 5 * 3?",
          reasoning: [],
        }),

        // Second assistant message
        createMockEvent({
          id: "assistant-2",
          role: "assistant",
          content: undefined,
          reasoning: [
            "The user is asking another math question",
            "I should use the calculator tool again",
          ],
        }),
        createMockEvent({
          id: "assistant-2",
          role: "assistant",
          content: "I'll calculate that too.",
          reasoning: [
            "The user is asking another math question",
            "I should use the calculator tool again",
          ],
        }),
        createMockEvent({
          id: "assistant-2",
          role: "assistant",
          content: "I'll calculate that too.",
          toolCalls: [
            {
              id: "call_calc_2",
              type: "function" as const,
              function: {
                name: "calculator",
                arguments: '{"expression": "5 * 3"}',
              },
            },
          ],
          reasoning: [
            "I need to call the calculator tool with the new expression",
          ],
        }),
        createMockEvent({
          id: "assistant-2",
          role: "assistant",
          content: "I'll calculate that too.",
          toolCalls: [
            {
              id: "call_calc_2",
              type: "function" as const,
              function: {
                name: "calculator",
                arguments: '{"expression": "5 * 3"}',
              },
            },
          ],
          reasoning: [
            "I need to call the calculator tool with the new expression",
          ],
        }),
      ];

      let eventIndex = 0;
      jest.mocked(mockStream[Symbol.asyncIterator]).mockReturnValue({
        next: async () => {
          if (eventIndex < events.length) {
            return { value: events[eventIndex++], done: false };
          }
          return { done: true, value: undefined };
        },
      } as AsyncIterableIterator<AgentResponse>);

      const results: LegacyComponentDecision[] = [];
      for await (const result of runAgentLoop(
        mockAgentClient,
        messages,
        createMockTools(),
      )) {
        results.push(result);
      }

      expect(results).toMatchSnapshot();
    });
  });

  describe("edge cases", () => {
    it("should handle empty message content", async () => {
      const messages: ThreadMessage[] = [];

      const events = [
        createMockEvent({
          id: "assistant-1",
          role: "assistant",
          content: "",
          reasoning: [],
        }),
      ];

      let eventIndex = 0;
      jest.mocked(mockStream[Symbol.asyncIterator]).mockReturnValue({
        next: async () => {
          if (eventIndex < events.length) {
            return { value: events[eventIndex++], done: false };
          }
          return { done: true, value: undefined };
        },
      } as AsyncIterableIterator<AgentResponse>);

      const results: LegacyComponentDecision[] = [];
      for await (const result of runAgentLoop(
        mockAgentClient,
        messages,
        createMockTools(),
      )) {
        results.push(result);
      }

      expect(results).toMatchSnapshot();
    });

    it("should handle assistant message without tool calls", async () => {
      const messages: ThreadMessage[] = [];

      const events = [
        createMockEvent({
          id: "assistant-1",
          role: "assistant",
          content: "This is a simple response.",
          reasoning: [],
        }),
      ];

      let eventIndex = 0;
      jest.mocked(mockStream[Symbol.asyncIterator]).mockReturnValue({
        next: async () => {
          if (eventIndex < events.length) {
            return { value: events[eventIndex++], done: false };
          }
          return { done: true, value: undefined };
        },
      } as AsyncIterableIterator<AgentResponse>);

      const results: LegacyComponentDecision[] = [];
      for await (const result of runAgentLoop(
        mockAgentClient,
        messages,
        createMockTools(),
      )) {
        results.push(result);
      }

      expect(results).toMatchSnapshot();
    });

    it("should handle system message", async () => {
      const messages: ThreadMessage[] = [];

      const events = [
        createMockEvent({
          id: "system-1",
          role: "system",
          content: "You are a helpful assistant.",
          reasoning: [],
        }),
      ];

      let eventIndex = 0;
      jest.mocked(mockStream[Symbol.asyncIterator]).mockReturnValue({
        next: async () => {
          if (eventIndex < events.length) {
            return { value: events[eventIndex++], done: false };
          }
          return { done: true, value: undefined };
        },
      } as AsyncIterableIterator<AgentResponse>);

      const results: LegacyComponentDecision[] = [];
      for await (const result of runAgentLoop(
        mockAgentClient,
        messages,
        createMockTools(),
      )) {
        results.push(result);
      }

      expect(results).toMatchSnapshot();
    });
  });
});
