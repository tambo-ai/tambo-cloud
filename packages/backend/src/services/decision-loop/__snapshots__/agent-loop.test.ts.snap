// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`runAgentLoop complete conversation flow should handle full conversation with assistant, tool call, tool result, and follow-up 1`] = `
[
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-1",
    "message": "",
    "props": null,
    "reasoning": [
      "The user is asking a math question",
      "I should use a calculator tool",
    ],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": undefined,
    "toolCallRequest": undefined,
  },
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-1",
    "message": "I'll calculate that for you.",
    "props": null,
    "reasoning": [
      "The user is asking a math question",
      "I should use a calculator tool",
    ],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": undefined,
    "toolCallRequest": undefined,
  },
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-1",
    "message": "I'll calculate that for you.",
    "props": null,
    "reasoning": [
      "I need to call the calculator tool with the expression",
    ],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": "call_calc_1",
    "toolCallRequest": {
      "parameters": [
        {
          "parameterName": "expression",
          "parameterValue": "2 + 2",
        },
      ],
      "toolName": "calculator",
    },
  },
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "tool-1",
    "message": "4",
    "props": null,
    "reasoning": [],
    "role": "tool",
    "statusMessage": "",
    "toolCallId": "call_calc_1",
    "toolCallRequest": undefined,
  },
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "user-2",
    "message": "What about 5 * 3?",
    "props": null,
    "reasoning": [],
    "role": "user",
    "statusMessage": "",
    "toolCallId": undefined,
    "toolCallRequest": undefined,
  },
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-2",
    "message": "",
    "props": null,
    "reasoning": [
      "The user is asking another math question",
      "I should use the calculator tool again",
    ],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": undefined,
    "toolCallRequest": undefined,
  },
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-2",
    "message": "I'll calculate that too.",
    "props": null,
    "reasoning": [
      "The user is asking another math question",
      "I should use the calculator tool again",
    ],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": undefined,
    "toolCallRequest": undefined,
  },
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-2",
    "message": "I'll calculate that too.",
    "props": null,
    "reasoning": [
      "I need to call the calculator tool with the new expression",
    ],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": "call_calc_2",
    "toolCallRequest": {
      "parameters": [
        {
          "parameterName": "expression",
          "parameterValue": "5 * 3",
        },
      ],
      "toolName": "calculator",
    },
  },
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-2",
    "message": "I'll calculate that too.",
    "props": null,
    "reasoning": [
      "I need to call the calculator tool with the new expression",
    ],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": "call_calc_2",
    "toolCallRequest": {
      "parameters": [
        {
          "parameterName": "expression",
          "parameterValue": "5 * 3",
        },
      ],
      "toolName": "calculator",
    },
  },
]
`;

exports[`runAgentLoop edge cases should handle assistant message without tool calls 1`] = `
[
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-1",
    "message": "This is a simple response.",
    "props": null,
    "reasoning": [],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": undefined,
    "toolCallRequest": undefined,
  },
]
`;

exports[`runAgentLoop edge cases should handle empty message content 1`] = `
[
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-1",
    "message": "",
    "props": null,
    "reasoning": [],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": undefined,
    "toolCallRequest": undefined,
  },
]
`;

exports[`runAgentLoop edge cases should handle system message 1`] = `
[
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "system-1",
    "message": "You are a helpful assistant.",
    "props": null,
    "reasoning": [],
    "role": "system",
    "statusMessage": "",
    "toolCallId": undefined,
    "toolCallRequest": undefined,
  },
]
`;

exports[`runAgentLoop message event sequences should handle complete assistant message sequence 1`] = `
[
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-1",
    "message": "",
    "props": null,
    "reasoning": [
      "I need to respond to the user's greeting",
    ],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": undefined,
    "toolCallRequest": undefined,
  },
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-1",
    "message": "Hello! I'm doing well, thank you for asking. ",
    "props": null,
    "reasoning": [
      "I should be friendly and respond to their greeting",
    ],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": undefined,
    "toolCallRequest": undefined,
  },
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-1",
    "message": "Hello! I'm doing well, thank you for asking. How can I help you today?",
    "props": null,
    "reasoning": [
      "I should also offer to help them",
    ],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": undefined,
    "toolCallRequest": undefined,
  },
]
`;

exports[`runAgentLoop message event sequences should handle user message sequence 1`] = `
[
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "user-1",
    "message": "",
    "props": null,
    "reasoning": [],
    "role": "user",
    "statusMessage": "",
    "toolCallId": undefined,
    "toolCallRequest": undefined,
  },
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "user-1",
    "message": "What's the weather like?",
    "props": null,
    "reasoning": [],
    "role": "user",
    "statusMessage": "",
    "toolCallId": undefined,
    "toolCallRequest": undefined,
  },
]
`;

exports[`runAgentLoop reasoning sequences should handle assistant message with reasoning 1`] = `
[
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-1",
    "message": "",
    "props": null,
    "reasoning": [
      "The user is asking about the weather",
      "I should check the current weather conditions",
      "I'll need to use a weather tool",
    ],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": undefined,
    "toolCallRequest": undefined,
  },
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-1",
    "message": "Let me check the weather for you.",
    "props": null,
    "reasoning": [
      "The user is asking about the weather",
      "I should check the current weather conditions",
      "I'll need to use a weather tool",
      "I should let them know I'm looking it up",
    ],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": undefined,
    "toolCallRequest": undefined,
  },
]
`;

exports[`runAgentLoop reasoning sequences should handle reasoning without content 1`] = `
[
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-1",
    "message": "",
    "props": null,
    "reasoning": [
      "I need to think about this problem",
      "This is a complex question that requires analysis",
      "Let me break it down step by step",
    ],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": undefined,
    "toolCallRequest": undefined,
  },
]
`;

exports[`runAgentLoop tool call event sequences should handle complete tool call sequence 1`] = `
[
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-1",
    "message": "I'll help you with that.",
    "props": null,
    "reasoning": [
      "I should use a tool to get the information",
    ],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": undefined,
    "toolCallRequest": undefined,
  },
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-1",
    "message": "I'll help you with that.",
    "props": null,
    "reasoning": [
      "I need to call the test_tool",
    ],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": "call_123",
    "toolCallRequest": {
      "parameters": [],
      "toolName": "test_tool",
    },
  },
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-1",
    "message": "I'll help you with that.",
    "props": null,
    "reasoning": [
      "I've prepared the tool call with parameters",
    ],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": "call_123",
    "toolCallRequest": {
      "parameters": [
        {
          "parameterName": "param1",
          "parameterValue": "value1",
        },
        {
          "parameterName": "param2",
          "parameterValue": 42,
        },
      ],
      "toolName": "test_tool",
    },
  },
]
`;

exports[`runAgentLoop tool call event sequences should handle tool call with invalid JSON arguments 1`] = `
[
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "assistant-1",
    "message": "I'll call a tool.",
    "props": null,
    "reasoning": [
      "I need to call the test_tool",
    ],
    "role": "assistant",
    "statusMessage": "",
    "toolCallId": "call_123",
    "toolCallRequest": undefined,
  },
]
`;

exports[`runAgentLoop tool call event sequences should handle tool result message 1`] = `
[
  {
    "completionStatusMessage": "",
    "componentName": null,
    "componentState": null,
    "id": "tool-1",
    "message": "Tool execution result",
    "props": null,
    "reasoning": [],
    "role": "tool",
    "statusMessage": "",
    "toolCallId": "call_123",
    "toolCallRequest": undefined,
  },
]
`;
